File: CSharp/Items/MSCE.IPrimWpf/PrimitiveToolWpf.cs
/*--------------------------------------------------------------------------------------+
|   $safeitemname$.cs
|
+--------------------------------------------------------------------------------------*/

#region Bentley Namespaces
using BCOM = Bentley.Interop.MicroStationDGN;
using BMI = Bentley.MstnPlatformNET.InteropServices;
using BMW = Bentley.MstnPlatformNET.WPF;
#endregion

namespace $rootnamespace$
{
    class $safeitemname$ : BCOM.IPrimitiveCommandEvents
    {
    //Use $safeitemname$UserControl.Property, $safeitemname$UserControl.Method(), or $safeitemname$UserControl.Field
    //to access public properties, methods, or fields in $safeitemname$.xaml.cs.

    #region ToolSettings
    private static $safeitemname$ $safeitemname$Command { get; set; }
private static BMW.ToolSettingsHost $safeitemname$ToolSettingsHost { get; set; }
private static $safeitemname$UC $safeitemname$UserControl { get; set; }

class MyToolSettings : BMW.ToolSettingsHost
{
    private $safeitemname$ m_command ;

    public MyToolSettings($safeitemname$ command)
    {
        m_command = command;

        $safeitemname$UserControl = new $safeitemname$UC();
        this.Content = $safeitemname$UserControl;
        this.Title = "$safeitemname$ Tool Settings";
    }
};
#endregion

public $safeitemname$()
{
    $safeitemname$ToolSettingsHost = new MyToolSettings(this);
}

#region IPrimitiveCommandEvents Members

public void Start()
{
    $safeitemname$ToolSettingsHost.Attach(Program.Addin);
}

public void Cleanup()
{
    if (null != $safeitemname$Command)
    {
        $safeitemname$ToolSettingsHost.Detach();
        $safeitemname$ToolSettingsHost.Dispose();
        $safeitemname$ToolSettingsHost = null;
        $safeitemname$Command = null;
    }
}

public void DataPoint(ref BCOM.Point3d Point, BCOM.View View)
{
}

public void Dynamics(ref BCOM.Point3d Point, BCOM.View View, BCOM.MsdDrawingMode DrawMode)
{
}

public void Keyin(string Keyin)
{
}

public void Reset()
{

    BMI.Utilities.ComApp.CommandState.StartDefaultCommand();
}

#endregion

/// <summary>
/// Open $safeitemname$ WPF user control.
/// </summary>
/// <param name="unparsed"></param>
public void Run(string unparsed = "")
{
    if (null == $safeitemname$Command)
    {
        $safeitemname$Command = new $safeitemname$();
        BMI.Utilities.ComApp.CommandState.StartPrimitive($safeitemname$Command, false);
    }
    else
        $safeitemname$ToolSettingsHost.Focus();
}
}
    }

File: Visual Basic/Items/MSCE.IPrimWpf/PrimitiveToolWpf.vb
'*--------------------------------------------------------------------------------------+
'   $safeitemname$.vb
'
'+--------------------------------------------------------------------------------------*/

#Region "Bentley Namespaces"
Imports BCOM = Bentley.Interop.MicroStationDGN
Imports BMI = Bentley.MstnPlatformNET.InteropServices
Imports BMW = Bentley.MstnPlatformNET.WPF
#End Region

Class $safeitemname$
    Implements BCOM.IPrimitiveCommandEvents

    'Use $safeitemname$UserControl.Property, $safeitemname$UserControl.Method(), Or $safeitemname$UserControl.Field
    'to access public properties, methods, or fields in $safeitemname$.xaml.vb.

#Region "ToolSettings"
    Private Shared Property $safeitemname$Command As $safeitemname$
    Private Shared Property $safeitemname$ToolSettingsHost As BMW.ToolSettingsHost
    Private Shared Property $safeitemname$UserControl As $safeitemname$UC

    Class MyToolSettings
        Inherits BMW.ToolSettingsHost

        Private Property m_command As $safeitemname$

        Public Sub New(ByVal command As $safeitemname$)
            m_command = command
            $safeitemname$UserControl = New $safeitemname$UC()
            Me.Content = $safeitemname$UserControl
            Me.Title = "$safeitemname$ Tool Settings"
        End Sub
    End Class
#End Region

#Region "IPrimitiveCommandEvents Members"

    Public Sub New()
        $safeitemname$ToolSettingsHost = New MyToolSettings(Me)
    End Sub

    Public Sub Start() Implements BCOM.IPrimitiveCommandEvents.Start
        $safeitemname$ToolSettingsHost.Attach(Program.Addin)
    End Sub

    Public Sub Cleanup() Implements BCOM.IPrimitiveCommandEvents.Cleanup
        If $safeitemname$Command IsNot Nothing Then
            $safeitemname$ToolSettingsHost.Detach()
            $safeitemname$ToolSettingsHost.Dispose()
            $safeitemname$ToolSettingsHost = Nothing
            $safeitemname$Command = Nothing
        End If
    End Sub

    Public Sub DataPoint(ByRef Point As BCOM.Point3d, ByVal View As BCOM.View) Implements BCOM.IPrimitiveCommandEvents.DataPoint
    End Sub

    Public Sub Dynamics(ByRef Point As BCOM.Point3d, ByVal View As BCOM.View, ByVal DrawMode As BCOM.MsdDrawingMode) Implements BCOM.IPrimitiveCommandEvents.Dynamics
    End Sub

    Public Sub Keyin(ByVal Keyin As String) Implements BCOM.IPrimitiveCommandEvents.Keyin
    End Sub

    Public Sub Reset() Implements BCOM.IPrimitiveCommandEvents.Reset

        BMI.Utilities.ComApp.CommandState.StartDefaultCommand()
    End Sub
#End Region

    Public Sub Run(ByVal Optional unparsed As String = "")
        If $safeitemname$Command Is Nothing Then
            $safeitemname$Command = New $safeitemname$()
            BMI.Utilities.ComApp.CommandState.StartPrimitive($safeitemname$Command, False)
        Else
            $safeitemname$ToolSettingsHost.Focus()
        End If
    End Sub
End Class